### PR TITLE
Improve search DSL and add selected filtering

### PR DESCRIPTION
**Pull request by huwshimi**
## Done
- Clean up DSL destructuring.
- Implement selected filtering.
- Fix DSL bugs.

## QA
- Visit /r/machines.
- Select some machines.
- In the search box search for `in:selected`, you should only see your selected machines.
- In the search box search for `in:!selected`, you should not see your selected machines.
- Check that you can search in other ways according to: https://maas.io/docs/interactive-search

## Fixes
Fixes: https://github.com/canonical-web-and-design/maas-squad/issues/1799.

### CODE REVIEW

File: ui/src/app/machines/filter-nodes.js [hunk 1]
```diff
@@ -54,15 +58,15 @@ const filterNodes = (nodes, search) => {
...
-        nodes.forEach(node => {
-          if (node.$selected && term.toLowerCase() === "selected") {
-            matched.push(node);
-          } else if (!node.$selected && term.toLowerCase() === "!selected") {
-            matched.push(node);
-          }
-        });
-        nodes = matched;
+        if (term.toLowerCase() === "selected") {
+          nodes = nodes.filter(({ system_id }) =>
```
Comment (hatched) [inline ui/src/app/machines/filter-nodes.js]: You're redefining the function arguments, typically this is a no-no. I'm surprised the linter didn't catch it...

I probably just don't have enough context but this seems odd...So for each `term` you're redefining `nodes` with a `filter` result? So any subsequent `term` will override any previous values. Then once `terms` is exhausted it just returns that value? It looks like you could possibly return the result from a `terms.filter` on line 60.

Or you an disregard all of this if I'm way off base 😄

File: ui/src/app/base/selectors/machine/machine.js [hunk 2]
```diff
@@ -69,12 +69,16 @@ machine.getBySystemId = (state, id) =>
...
+ * Gets the search terms and selected machines.
  * @param {Object} state - The redux state.
- * @param {Array} terms - The search terms.
- * @returns {Array} The search terms.
+ * @param {String} terms - The search string.
+ * @param {Array} selectedIDs - The selected machine ids.
+ * @returns {Array} The search terms and selected machines.
  */
-machine._getTerms = (state, terms) => terms;
+machine._getSearchParams = (state, terms, selectedIDs) => ({
```
Comment (hatched) [inline ui/src/app/base/selectors/machine/machine.js]: This is interesting way to curry arguments into selectors. You may prefer how we do it on the jaas-dashboard https://github.com/canonical-web-and-design/jaas-dashboard/blob/master/src/app/selectors.js#L448

File: ui/src/app/machines/search.js [hunk 3]
```diff
@@ -10,55 +10,39 @@ export const getCurrentFilters = search => {
...
-          filter.push(termItem);
-        }
-      });
-      // If the term contains the ending ')' then it's the last in the group.
-      if (term.includes(")")) {
-        processingFilter = null;
+  // Match filters with parens e.g. 'status:(new,deployed)'.
+  // Then: match filters without parens e.g. 'status:new,deployed' or 'status'
+  const filterMatchingRegex = search.matchAll(/(\b\w+:!?\([^)]+\))|(\w+\S+)/g);
+  Array.from(filterMatchingRegex).forEach(([group]) => {
```
Comment (hatched) [inline ui/src/app/machines/search.js]: You could use the spread operator to make this a bit shorter if you'd like.
```JavaScript
[...filterMatchingRegex]
```

**Comment by cristinadresch** [comment]
@hatched do you want to give it a go reviewing this one?

**Comment by hatched** [comment]
@cristinadresch I have already....and no updates have been made since my comments were submitted.

**Comment by cristinadresch** [comment]
@huwshimi spoke to Jeff, he said he didn't want to out code -1 because he doesn't have enough context. but if you are making any changes and this is not ready for another review, could you remove the labels? just so we know what it's happening in our standup.

cc @hatched

File: ui/src/app/machines/search.js (hunk 3 above)
Comment (huwshimi) [reply ui/src/app/machines/search.js]: Done.

File: ui/src/app/base/selectors/machine/machine.js (hunk 2 above)
Comment (huwshimi) [reply ui/src/app/base/selectors/machine/machine.js]: I decided against this approach as it makes using the selector more complicated (looks like you wrap the selector creator in `useMemo` before you use `useSelector` on the resulting function.

File: ui/src/app/machines/filter-nodes.js (hunk 1 above)
Comment (huwshimi) [reply ui/src/app/machines/filter-nodes.js]: Fixed (and a few bugs found along the way too).

**Comment by huwshimi** [comment]
This required significant reworking after @hatched's comments. I've just finished those so ready for another review.

@hatched your comments have been very helpful, I'd be happy with your review if you have time to do another one.

@lilyvidenova maybe you could QA this one to save @hatched having to get a MAAS set up. Would that be OK?

File: ui/src/app/base/selectors/machine/machine.js (hunk 2 above)
Comment (hatched) [reply ui/src/app/base/selectors/machine/machine.js]: Sure no problem, maybe we can convert the dashboard to use this approach instead. Something to investigate.

**Comment by lilyvidenova** [comment]
@huwshimi terribly sorry for still blocking this, however I can't get my maas-ui to see the internet through multipass. Ant kindly added the ./run back to the ui today, could you maybe include that bit into this branch so I can run it on my mac in order to QA it?

**Comment by huwshimi** [comment]
@lilyvidenova this has now been updated with the `./run` script.

**Comment by lilyvidenova** [comment]
Looks great, thanks @huwshimi

**Pull request merged**